pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,96 @@
+from datetime import timedelta
+import numbers
+from typing import (
+    Generic,
+    TypeVar,
+)
+
+import numpy as np
+import numpy.typing as npt
+
+from pandas._typing import (
+    IntervalCloseness,
+    Scalar,
+)
+
+_ScalarT = TypeVar("_ScalarT", bound=Scalar)
+
+class IntervalMixin:
+    @property
+    def closed_left(self) -> bool: ...
+    @property
+    def closed_right(self) -> bool: ...
+    @property
+    def open_left(self) -> bool: ...
+    @property
+    def open_right(self) -> bool: ...
+    @property
+    def mid(self) -> float: ...
+    @property
+    def length(self) -> float: ...
+    @property
+    def is_empty(self) -> bool: ...
+    def _check_closed_matches(self, other: IntervalMixin, name: str = ...) -> None: ...
+
+class Interval(IntervalMixin, Generic[_ScalarT]):
```

Comment:
Two points that are probably okay but might be worth pointing out. While I want `Interval` to be a `Generic`, it is itself part of `Scalar`:
- `Scalar` is now a mix of concrete types and one generic type
- `Interval` is declared to allow intervals where left and right are intervals themselves